wpilibj/src/main/java/edu/wpi/first/wpilibj/simulation/DifferentialDrivetrainSim.java

```diff
@@ -0,0 +1,171 @@
+
+  /**
+   * Create a SimDrivetrain.
+   *
+   * @param drivetrainPlant The {@link LinearSystem} representing the robot's drivetrain. This
+   *                        system can be created with {@link edu.wpi.first.wpilibj.system.plant.LinearSystemId#createDrivetrainVelocitySystem(DCMotor, double, double, double, double, double)}
+   *                        or {@link edu.wpi.first.wpilibj.system.plant.LinearSystemId#identifyDrivetrainSystem(double, double, double, double)}.
+   * @param kinematics      A {@link DifferentialDriveKinematics} object representing the
+   *                        differential drivetrain's kinematics.
+   */
+  public DifferentialDrivetrainSim(LinearSystem<N2, N2, N2> drivetrainPlant,
+                                   DifferentialDriveKinematics kinematics,
+                                   DCMotor leftGearbox, double leftGearing,
+                                   DCMotor rightGearbox, double rightGearing) {
+    this.m_plant = drivetrainPlant;
+    this.m_rb = kinematics.trackWidthMeters / 2.0;
+    this.m_leftMotor = leftGearbox;
+    this.m_rightMotor = rightGearbox;
+    this.m_leftGearing = leftGearing;
+    this.m_rightGearing = rightGearing;
+
+    m_x = new Matrix<>(new SimpleMatrix(10, 1));
+  }
+
+  public void setInputs(double leftVoltageVolts, double rightVoltageVolts) {
+    m_u = VecBuilder.fill(leftVoltageVolts, rightVoltageVolts);
+  }
+
+  @SuppressWarnings("LocalVariableName")
+  public void update(double dtSeconds) {
+
+    // Update state estimate with RK4
+    m_x = RungeKutta.rungeKutta(this::getDynamics, m_x, m_u, dtSeconds);
+  }
+
+  public double getState(State state) {
+    return m_x.get(state.value, 0);
+  }
+
+  public Matrix<N10, N1> getState() {
+    return m_x;
+  }
+
+  /**
+   * Get the estimated direction the robot is pointing. Note that this angle is
+   * counterclockwise-positive, while most gyros are clockwise positive.
+   */
+  public Rotation2d getHeading() {
+    return new Rotation2d(getState(State.kHeading));
+  }
+
+  public Pose2d getEstimatedPosition() {
+    return new Pose2d(m_x.get(0, 0),
+      m_x.get(1, 0),
+      new Rotation2d(m_x.get(2, 0)));
+  }
+
+  public double getCurrentDrawAmps() {
+    var loadIleft = m_leftMotor.getCurrent(
+        getState(State.kLeftVelocity) * m_leftGearing / m_rb,
+        m_u.get(0, 0)) * Math.signum(getState(State.kLeftVelocity));
+
+    var loadIright = m_rightMotor.getCurrent(
+        getState(State.kRightVelocity) * m_rightGearing / m_rb,
+        m_u.get(1, 0)) * Math.signum(getState(State.kRightVelocity));
+
+    return loadIleft + loadIright;
+  }
+
+  @SuppressWarnings({"DuplicatedCode", "LocalVariableName"})
+  protected Matrix<N10, N1> getDynamics(Matrix<N10, N1> x, Matrix<N2, N1> u) {
+    Matrix<N4, N2> B = new Matrix<>(new SimpleMatrix(4, 2));
+    B.getStorage().insertIntoThis(0, 0, m_plant.getB().getStorage());
+    B.getStorage().insertIntoThis(2, 0, new SimpleMatrix(2, 2));
```

Comment:
I'm not sure if this compils as is, but something like it would be better than what's there now:
```suggestion
    Matrix<N4, N2> B = new Matrix<>(Nat.N4(), Nat.N2());
    B.assignBlock(0, 0, m_plant.getB());
    B.assignBlock(2, 0, new Matrix<>(Nat.N2(), Nat.N2()));
```